### DISCUSSION TITLE
can't set output resolution

### DISCUSSION THREAD

**Discussion by derchipo** [letmaik/pyvirtualcam#41] [Q&A]
Thanks vor this lib. It works fine for me. But there ist one Problem

After multiple tests it is not possible to set the outputresolution.

cam = pyvirtualcam.Camera(width=1920, height=1080, fps=18, print_fps=True)
cam = pyvirtualcam.Camera(width=1280, height=720, fps=18, print_fps=True)
cam = pyvirtualcam.Camera(width=768, height=432, fps=18, print_fps=True)
cam = pyvirtualcam.Camera(width=480, height=270, fps=18, print_fps=True)
and all with 
cam = pyvirtualcam.Camera(480, 270, fps=18, print_fps=True)
all of these inits procuces the same Output: 1920x1080 with very large img sizes
How can I reduce the Quality to get a better stream with bad uploadspeed.
Or how can i set the default output resolution.

**Comment by letmaik** [comment]
Could you try again with the latest version? Also, could you make sure to first start the Python script and only then use the camera in other apps?

**Comment by derchipo** [comment]
Hi,

 

the latest verion is not working form me.

 

When i have instelled OBS Studio pyvirtualcam in version 0.5 uses the OBS-Virtual-Camera and not the OBS-Camera wich i registerd.

In Version 0.34 it works but there i have the problem setting the outputreolution

 

 

 

Von: Maik Riechert ***@***.***> 
Gesendet: Mittwoch, 10. März 2021 21:46
An: letmaik/pyvirtualcam ***@***.***>
Cc: derchipo ***@***.***>; Author ***@***.***>
Betreff: Re: [letmaik/pyvirtualcam] can't set output resolution (#14)

 

Could you try again with the latest version? Also, could you make sure to first start the Python script and only then use the camera in other apps?

**Comment by letmaik** [comment]
Can you check the README? There have been some changes in newer versions.

**Comment by derchipo** [comment]
Yes,

 

i saw that in the new version the lib select the first avalable device. This crashes my script when i habeOBS-Studio installed. Maybee it’s a bug ? 

 

Von: Maik Riechert ***@***.***> 
Gesendet: Dienstag, 16. März 2021 18:37
An: letmaik/pyvirtualcam ***@***.***>
Cc: derchipo ***@***.***>; Author ***@***.***>
Betreff: Re: [letmaik/pyvirtualcam] can't set output resolution (#14)

 

Can you check the README? There have been some changes in newer versions.

**Comment by justaquickhack** [comment]
Same topic, different backend.

I do not have this issue when using the OBS Virtual Camera, that sets the output correctly, but when switching to Unity Video Capture, the output is always 1440x900 and 30 fps, regardless of the parameters.  I am using pyvirtualcam-0.8.0 and UnityCapture current release.

For example:  cam = pyvirtualcam.Camera(width=640, height=480, fps=25, device='Unity Video Capture', backend='unitycapture')

I even tried it with your Usage example code in the README, with the same result.
Is there something that I am missing?

Thank you

**Reply by letmaik** [reply]
Input resolution != output resolution. The consumer (browser, skype etc.) selects the output resolution from multiple supported ones. The resolution you set in pyvirtualcam only affects the input resolution.

**Reply by justaquickhack** [reply]
Please forgive my ignorance.  If I set the output as above, with OBS Virtual Camera, it displays in VLC at the resolution I specified, if I use UnityCapture, it is always 1440x900.  Why would VLC match the OBS and not the UnityCapture, using the same exact code, except for the backend?  I thought that pyvirtualcam was setting the resolution of the virtual camera device, so that applications would be presented with a camera at that resolution.

**Reply by letmaik** [reply]
There are three parts to virtual cameras. There's the actual device driver (the thing you have to install separately), there's the part that sends frames to the driver (pyvirtualcam), and there's the app that connects to the driver as a receiver (VLC). pyvirtualcam tells the driver which input resolution the frames are that it will send to it. The driver in turn tells the operating system which output resolutions it supports, and this is typically independent of the input resolutions. A mismatch in input/output resolution means the driver has to resize the frames. The app that connects to the driver to receive frames (VLC) queries available output resolutions and picks one based on some preference or UI setting by the user. For example, in Zoom, if I don't enable HD, then Zoom prefers 640x480, while with HD it picks 1280x720. The reason why you see a difference in VLC between OBS and UnityCapture is likely that by default VLC picks the first advertised output resolution, and I assume OBS and UnityCapture advertise them in different orderings. You should be able to influence that choice by specifying your desired resolution. When I go to Open Media -> Capture Device, I see a "Video size" input field. This will probably allow you choose the right resolution.

**Reply by justaquickhack** [reply]
Letmaik,

Thank you for taking the time to provide such a concise explanation!  I understand it now.
I really appreciate your work and understanding.

Vielen Dank!
